### ISSUE TITLE
Feature request: overlaymode only in certain tokens.

### ISSUE THREAD

**Issue by Janpot** [codemirror/CodeMirror#5029]
I'd like to be able to create an overlay mode, but one that only kicks in on certain tokens in the base mode. I've been digging through the code but it seems like there is no way to achieve this behaviour right now.
It could either be done by communicating the current token of the base mode to the `token` method of the overlay mode. Or by specifying an array of tokens when the `overlayMode` function is called.
Would something like this be something you'd consider supporting?

**Comment by marijnh** [comment]
Are you talking about the mode overlay addon, or about overlays in terms of `addOverlay`/`removeOverlay`?

In the second case, there's an implementation in [this branch](https://github.com/codemirror/CodeMirror/tree/overlay_baseToken), but I'm still waiting for feedback from a customer before I merge that.

**Comment by Janpot** [comment]
I was talking about the addon, but looks like I could work with `addOverlay` too.
More specific to my use case, I'm making a _visible whitespace_ mode, but I want it to apply only on string and string-2 of the base mode. (in retrospect, it seems like `addOverlay` is actually a better fit than the addon here)
Looking at the code in the branch you posted, it seems like I could use it.

**Comment by marijnh** [comment]
In that case, a wrapping mode sounds more appropriate. I.e. don't use the overlay addon, just write a wrapping mode from scratch which calls through to the inner mode's `token` method, and then only adds its own styling when that is one of the token types you're interested in.

**Comment by Janpot** [comment]
So just so I understand you correctly, imagine I'm creating an editor with a select box that selects `mode` and checkbox 'visible whitespace'. Anytime these change I'd do something like
```js
function onChange () {
  if (isVisibleWhiteSpaceChecked) {
    cm.setOption('mode', visibleWhiteSpace(mode, { restrictTo: ['string'] }));
  } else {
    cm.setOption('mode', mode);
  }
}
```

But I had more something like this in mind:

```js
function onSelect (mode) {
  cm.setOption('mode', mode);
}
function onCheck (visibleWhiteSpace) {
  if (visibleWhiteSpace) {
    // this mode could be made with the baseToken branch
    cm.addOverlay({ name: 'visible-whitespace', restrictTo: ['string'] });
  } else {
    cm.removeOverlay('visible-whitespace');
  }
}
```
Which, to me, has the benefit that I can distribute it as just any other mode, and it could also be used as a base mode.

Anyway, it's not _that_ big of a deal, just thought it was a nicer use of API.

**Comment by marijnh** [comment]
I've merged the `baseToken` branch in 79d266e6 (see the manual.html change in that patch), which should allow you to do the `addOverlay` thing.

**Comment by Janpot** [comment]
Ok, very cool. Will have to dig a bit deeper later but I have a hunch there's an off-by-one in there somewhere, I seem to get some tokens with size `0` and type of the previous token where I expect a token of a single character.
Will see if I can cook up a reproduction when I have some more time.
